tensorflow/contrib/tensorrt/kernels/trt_engine_op.cc

```diff
@@ -30,85 +37,300 @@ using IRuntime = nvinfer1::IRuntime;
-      LOG(FATAL) << "Can't find device allocator for gpu device "
-                 << device->name();
-    }
-    allocator_ = std::make_shared<TRTDeviceAllocator>(dev_allocator);
-    infer->setGpuAllocator(allocator_.get());
-#endif
-    trt_engine_ptr_.reset(infer->deserializeCudaEngine(
-        serialized_engine_.c_str(), serialized_engine_.size(),
-        PluginFactoryTensorRT::GetInstance()));
-    trt_execution_context_ptr_.reset(trt_engine_ptr_->createExecutionContext());
-    // Runtime is safe to delete after engine creation
-    infer->destroy();
-    serialized_engine_.clear();
-  }
-  int num_binding = context->num_inputs() + context->num_outputs();
+void TRTEngineOp::ComputeAsync(tensorflow::OpKernelContext* ctx,
+                               tensorflow::AsyncOpKernel::DoneCallback done) {
+  auto ah = new AsyncHelper(done);
+  tensorflow::core::ScopedUnref SC(ah);
+  if (calibration_mode) {
+    auto TRT_RM = tensorflow::tensorrt::TRTResourceManager::instance();
+    auto res_mgr = TRT_RM->getManager("TRTCalibration");
+    tensorflow::tensorrt::TRTCalibrationResource* calib_res = nullptr;
+    auto status = res_mgr->LookupOrCreate(
+        funcdef_name_, "Calibrator", &calib_res,
+        {[ctx, this](tensorflow::tensorrt::TRTCalibrationResource** cr)
+             -> tensorflow::Status {
+          return this->AllocateCalibrationResources(ctx, cr);
+        }});
+    if (!status.ok()) {
+      ctx->SetStatus(status);
+      return;
+    }
+    ExecuteNativeSegment(ctx, ah);
+    int num_inputs = ctx->num_inputs();
+    // Pass input data to calibrator
+    std::unordered_map<string, void*> input_data;
+    for (int i = 0; i < num_inputs; i++) {
+      const Tensor& t = ctx->input(i);
+      void* data_address = GetTensorAddress(&t);
+      const auto device_tensor = dev_tensors_.at(i).AccessTensor(ctx);
+      CHECK_EQ(t.TotalBytes(),
+               device_tensor->TotalBytes());  // use the tensor so FW keeps it
+      input_data.emplace(StrCat("InputPH_", i), data_address);
+    }
+    VLOG(2) << "Filled map for sending";
+    // copied from cuda_kernel_helper since it seems only valid in *.cu.cc files
+    const cudaStream_t* stream = CHECK_NOTNULL(
+        reinterpret_cast<const cudaStream_t*>(ctx->op_device_context()
+                                                  ->stream()
+                                                  ->implementation()
+                                                  ->CudaStreamMemberHack()));
+    ah->Ref();  // Increment count for calculating calibration data
+    calib_res->calibrator_->setBatch(input_data, *stream, ah);
+    VLOG(2) << "Passed calibration data";
+    return;
+  }
+  int num_binding = ctx->num_inputs() + ctx->num_outputs();
   std::vector<void*> buffers(num_binding);
 
   size_t binding_index;
-  int num_batch = 0;
-  for (int i = 0; i < context->num_inputs(); i++) {
+  int num_batch = ctx->input(0).shape().dim_size(0);
+  int smallest_engine = 0;
```

Comment:
Please make a new method maybe called 'FindOrCreateEngine' to encapsulate these code from this line to line 267.